File: server/src/L2.Studio.Context/Entities/AssetRelease.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace L2.Studio.Context.Entities;

[Table("asset_releases")]
public sealed class AssetRelease
{
    [Key, Column("id"), DatabaseGenerated(DatabaseGeneratedOption.None)]
    public Guid Id { get; set; }
    [Column("game_version"), MaxLength(32)]
    public string GameVersion { get; set; } = "interlude";
    [Column("name"), MaxLength(120)]
    public required string Name { get; set; }
    [Column("notes"), MaxLength(4000)]
    public string? Notes { get; set; }
    [Column("status"), MaxLength(32)]
    public string Status { get; set; } = "draft";
    [Column("snapshot_hash"), MaxLength(64)]
    public required string SnapshotHash { get; set; }
    [Column("validation_status"), MaxLength(32)]
    public string ValidationStatus { get; set; } = "not_validated";
    [Column("validation_issues_json")]
    public string ValidationIssuesJson { get; set; } = "[]";
    [Column("validated_snapshot_hash"), MaxLength(64)]
    public string? ValidatedSnapshotHash { get; set; }
    [Column("validation_requested_at")]
    public DateTimeOffset? ValidationRequestedAt { get; set; }
    [Column("validated_at")]
    public DateTimeOffset? ValidatedAt { get; set; }
    [Column("manifest_path"), MaxLength(1024)]
    public string? ManifestPath { get; set; }
    [Column("manifest_hash"), MaxLength(64)]
    public string? ManifestHash { get; set; }
    [Column("login_scene_file_id")]
    public long? LoginSceneFileId { get; set; }
    [Column("login_camera_sequence"), MaxLength(256)]
    public string? LoginCameraSequence { get; set; }
    [Column("login_music_file_id")]
    public long? LoginMusicFileId { get; set; }
    [Column("primary_logo_file_id")]
    public long? PrimaryLogoFileId { get; set; }
    [Column("version_logo_file_id")]
    public long? VersionLogoFileId { get; set; }
    [Column("loading_artwork_file_id")]
    public long? LoadingArtworkFileId { get; set; }
    [Column("character_selection_scene_file_id")]
    public long? CharacterSelectionSceneFileId { get; set; }
    [Column("character_selection_camera_sequence"), MaxLength(256)]
    public string? CharacterSelectionCameraSequence { get; set; }
    [Column("created_at")]
    public DateTimeOffset CreatedAt { get; set; }
    [Column("updated_at")]
    public DateTimeOffset UpdatedAt { get; set; }
    [Column("published_at")]
    public DateTimeOffset? PublishedAt { get; set; }
    [Column("retired_at")]
    public DateTimeOffset? RetiredAt { get; set; }
    public ICollection<AssetReleaseArtifact> Artifacts { get; set; } = [];
    public ICollection<AssetReleaseEvent> Events { get; set; } = [];
}

File: web/app/types/models/asset-release.ts
export type AssetReleaseStatus = 'draft' | 'published' | 'active' | 'retired'
export type AssetReleaseValidationStatus =
  | 'not_validated'
  | 'queued'
  | 'running'
  | 'valid'
  | 'invalid'

export interface AssetReleaseSummary {
  id: string
  name: string
  notes: string | null
  status: AssetReleaseStatus
  validationStatus: AssetReleaseValidationStatus
  snapshotHash: string
  rootArtifactCount: number
  artifactCount: number
  sizeBytes: number
  manifestPath: string | null
  manifestHash: string | null
  createdAt: string
  updatedAt: string
  publishedAt: string | null
  retiredAt: string | null
  isActive: boolean
  isDesired: boolean
}

export interface AssetReleaseEntrypoints {
  loginSceneFileId: number | null
  loginScenePath: string | null
  loginCameraSequence: string | null
  loginMusicFileId: number | null
  loginMusicPath: string | null
  primaryLogoFileId: number | null
  primaryLogoPath: string | null
  versionLogoFileId: number | null
  versionLogoPath: string | null
  loadingArtworkFileId: number | null
  loadingArtworkPath: string | null
  characterSelectionSceneFileId: number | null
  characterSelectionScenePath: string | null
  characterSelectionCameraSequence: string | null
}

export interface AssetReleaseValidationIssue {
  code: string
  field: string | null
  message: string
}

export interface AssetReleaseArtifact {
  artifactId: string
  kind: string
  sourceKey: string
  buildFingerprint: string
  integrityStatus: string
  sizeBytes: number
  isRoot: boolean
}

export interface AssetReleaseEvent {
  id: number
  action: string
  occurredAt: string
}

export interface AssetReleaseDetail {
  release: AssetReleaseSummary
  entrypoints: AssetReleaseEntrypoints
  validationIssues: AssetReleaseValidationIssue[]
  artifacts: AssetReleaseArtifact[]
  events: AssetReleaseEvent[]
  pointerStatus: string
  pointerError: string | null
}

export interface AssetReleasePage {
  items: AssetReleaseSummary[]
  total: number
  page: number
  pageSize: number
}

export interface AssetReleaseResourceOption {
  fileId: number
  artifactId: string
  kind: string
  sourceKey: string
  label: string
  publicPath: string
  mediaType: string
  cameraSequences: string[]
}

export interface AssetReleaseResourcePage {
  items: AssetReleaseResourceOption[]
  total: number
  page: number
  pageSize: number
}
